frame/dex-router/src/lib.rs

```diff
@@ -0,0 +1,264 @@
+			  // bit
+		/// The maximum hops in the route.
+		#[pallet::constant]
+		type MaxHopsInRoute: Get<u32> + TypeInfo;
+		type PoolId: FullCodec
+			+ Default
+			+ TypeInfo
+			+ Eq
+			+ PartialEq
+			+ Ord
+			+ Copy
+			+ Debug
+			+ CheckedAdd
+			+ Zero
+			+ One;
+		type PoolTokenIndex: Copy + Debug + Eq + Into<u32> + From<u8>;
+		type StableSwapDex: CurveAmm<
+			AssetId = Self::AssetId,
+			Balance = Self::Balance,
+			AccountId = Self::AccountId,
+			PoolTokenIndex = Self::PoolTokenIndex,
+			PoolId = Self::PoolId,
+		>;
+		type ConstantProductDex: CurveAmm<
+			AssetId = Self::AssetId,
+			Balance = Self::Balance,
+			AccountId = Self::AccountId,
+			PoolTokenIndex = Self::PoolTokenIndex,
+			PoolId = Self::PoolId,
+		>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub type DexRoutes<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AssetId,
+		Blake2_128Concat,
+		T::AssetId,
+		DexRoute<T::PoolId, T::MaxHopsInRoute>,
+		OptionQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Number of hops in route exceeded maximum limit.
+		MaxHopsExceeded,
+		/// A Pool provided as part of route does not exist.
+		PoolDoesNotExist,
+		/// For given asset pair no route found.
+		NoRouteFound,
+		/// Some error occured while performing exchange.
+		ExchangeError,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RouteAdded {
+			who: T::AccountId,
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+		RouteDeleted {
+			who: T::AccountId,
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+		RouteUpdated {
+			who: T::AccountId,
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			old_route: Vec<DexRouteNode<T::PoolId>>,
+			updated_route: Vec<DexRouteNode<T::PoolId>>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> DexRouter<T::AccountId, T::AssetId, T::PoolId, T::Balance> for Pallet<T> {
+		fn update_route(
+			who: &T::AccountId,
+			asset_pair: CurrencyPair<T::AssetId>,
+			route: Option<Vec<DexRouteNode<T::PoolId>>>,
+		) -> Result<(), DispatchError> {
+			let k1 = asset_pair.base;
+			let k2 = asset_pair.quote;
+			match route {
+				Some(route_vec) => {
+					for r in &route_vec {
+						match r {
+							DexRouteNode::Curve(pool_id) => {
+								if !T::StableSwapDex::pool_exists(*pool_id) {
+									return Err(Error::<T>::PoolDoesNotExist.into())
+								}
+							},
+							DexRouteNode::Uniswap(pool_id) => {
+								if !T::ConstantProductDex::pool_exists(*pool_id) {
+									return Err(Error::<T>::PoolDoesNotExist.into())
+								}
+							},
+						}
+					}
+					let bounded_route =
+						route_vec.clone().try_into().map_err(|_| Error::<T>::MaxHopsExceeded)?;
```

Comment:
Please use BoundedVec as input as here you check if routes are valid even if you exceed the max hops.